Add io.ascii.get_read_trace() that shows how table got read

taldcroft opened the pull request:
I've always wanted to have some traceability into what readers got called and how when reading an ASCII table.  This implements that.  For example, without the second commit, which is (strictly speaking) off-topic, one sees there is a problem in that the fast readers are generally not getting called when guessing.  Note the ` "TypeError: __cinit__() got an unexpected keyword argument 'strict_names'"`.
```
In [1]: from astropy.io.ascii.ui import READ_TRIES

In [2]: from astropy.io import ascii

In [3]: b = ascii.read(['a b', '1 2'])

In [4]: READ_TRIES
Out[4]: 
[{'kwargs': {'Reader': astropy.io.ascii.ecsv.Ecsv,
   'fill_values': [('', '0')],
   'strict_names': True},
  'status': 'InconsistentTableError: ECSV header line like "# %ECSV <version>" not found as first line.  This is required for a ECSV file.'},

 {'kwargs': {'Reader': astropy.io.ascii.fixedwidth.FixedWidthTwoLine,
   'fill_values': [('', '0')],
   'strict_names': True},
  'status': 'InconsistentTableError: Position line should only contain delimiters and one other character, e.g. "--- ------- ---".'},

 {'kwargs': {'Reader': astropy.io.ascii.fastbasic.FastBasic,
   'fill_values': [('', '0')],
   'strict_names': True},
  'status': "TypeError: __cinit__() got an unexpected keyword argument 'strict_names'"},

 {'kwargs': {'Reader': astropy.io.ascii.basic.Basic,
   'fill_values': [('', '0')],
   'strict_names': True},
  'status': 'Success (guessing)'}]
```

Actually, there was *another* bug, which is that in the guess list the `Basic` reader was always before the `FastBasic` reader because of the default `fill_values` kwarg which is inserted by `ui.read()`.   So this PR fixes that as well.

In fact, once both commits in this PR are in place, then a third bug appears in that the fast reader doesn't handle a no-header table correctly when guessing and `names` is supplied.  More on that tomorrow.

hamogu commented:
This is  good thing. We should have done that long ago, because debugging why this reader and not that reader tried to read a specific file has been at the root of several issues already. The fast that we constantly move move around the order of the Readers on the guessing list speaks to that.
Once we start to introduce more heuristics as suggested in #2881 it will get even more messy and even harder to debug, so we definitely should have this traceability.
I did not try out this PR yet, but I'm totally going to make use of it for future io.ascii development.

taldcroft commented:
OK I'll move forward on this.  Maybe have the actual list be internal with a documented public function `get_read_trace()` that returns a copy of it.

hamogu commented:
Not sure this even needs a function. Very light documentation is enough. I don't think the average user needs to know about it (hide it deep in the docs). It's more a think we would ask for when we get a bug report.

embray commented:
I think it's fine to document this in the changelog.  I think it's useful to know about, but agree it's mostly just for debugging purposes.
I still sort of feel like this is just what `log.debug` is for though.  But if you think this is more useful...

taldcroft commented:
@hamogu - ready for final review.

@embray - I thought seriously about `log.debug`, but in the end got worried that there are cases where the output of interest (the read trace) could get buried in other debug output.  In practice right now I think `log.debug` is lightly used, but it might not stay that way.

taldcroft commented on astropy/io/ascii/ui.py:
```diff
@@ -412,6 +437,7 @@ def _get_guess_kwargs_list(read_kwargs):
     # FixedWidthTwoLine would also be read by Basic, so it needs to come first.
     if len(read_kwargs) > 0:
         for reader in [fixedwidth.FixedWidthTwoLine,
+                       fastbasic.FastBasic,
```
@hamogu - in practice `len(read_kwargs) > 0` is almost always true because of the default fill values.  Therefore this list should match the list below, hence the addition of `FastBasic` here.

hamogu commented:
This still fails the `space_delim_no_header.dat` table. Was that intentional reordering of the Readers that did that (i.e. did you do that because the fast readers are now earlier in the lest and that benefit outweighs the fact that the NoHeader situation is not read correctly)?

embray commented:
> In practice right now I think log.debug is lightly used, but it might not stay that way.

Of course, it *will* stay that way if no one uses the debug log for fear of it being too cluttered :)  The log can also be filtered by module, and I think even down to the function level, so you can just set it to log only messages from functions and modules you care about.

But like I said, whatever you think is most convenient.

taldcroft commented:
I was biting off too much with this PR and including (partial) fixes unrelated to `get_read_trace()`.  I've reverted those changes and rebased, and hopefully all tests pass now.

hamogu commented on docs/io/ascii/read.rst:
```diff
@@ -298,6 +303,15 @@ Guessing can be disabled in two ways::
+something isn't working as expected, use the
+`~astropy.io.ascii.get_read_trace()` function.  This returns a traceback of the
+attempted read formats for the last call to `~astropy.io.ascii.read()` where
+guessing was enabled.
```
It seems that this gets written even is guessing is *not* enabled?
https://github.com/astropy/astropy/pull/3688/files#diff-e6b2d1f1997a9b7e467819e2c98be409R251

taldcroft replied:
Good point!

hamogu commented:
Only one small comment. Otherwise this is good to go. (Travis failure seems unrelated.)
